DataTables, Evolve calc, Candies

mackhankins opened the pull request:
![pokenurse](https://cloud.githubusercontent.com/assets/2184080/17367109/3569da78-5954-11e6-989f-03b50cf22d9c.gif)
- DataTables (sorting and search)
- Child Tables (sorting and checkall)
- Automatically calculating possible evolves
- Candies

**Bugs**
- ~~I think when you click the checkall it tries to evolve / transfer that checkbox~~
- Modal selector problems (sub level objects)
- ~~Nidoran Males and Females are not loading correctly.~~

@vinnymac I can't get the modal selector to work.  You'll see the problem in the prep_display function or logging `pokemons.pokemon`.  Somehow on the prep_display function I wanted to send `d` to show_modal with no joy.  I also dropped the images in the table because they take up so much real estate.  I tried with smaller width / height, but so much whitespace.  Maybe we need a tighter sprite?

vinnymac committed:
display pokemon name in detail modal

vinnymac committed:
use anchor, style, and update title with name

vinnymac committed:
add some more info to the detail modal, style

vinnymac committed:
fix sorting breaking detail modal, add CP progress based on a fake max

mackhankins committed:
DataTables, Evolve calc, Candies

mackhankins committed:
remove extra th

mackhankins commented:
In main.js around line 248

```js
console.log(combinedPokemonList['Nidoran Male'])
```

Results
```bash
{ pokemon_id: 32,
  name: 'Nidoran Male',
  count: 2,
  pokes:
   [ { cp: 457,
       creation_time_ms: '1469933631195',
       deployed: false,
       id: '3593382990786328140',
       attack: 12,
       defense: 13,
       stamina: 11,
       iv: 80,
       pokemon_id: 32,
       name: 'Nidoran Male',
       nickname: 'Nidoran Male',
       favorite: -0 },
     { cp: 165,
       creation_time_ms: '1470191178192',
       deployed: false,
       id: '12560005815836317887',
       attack: 11,
       defense: 3,
       stamina: 5,
       iv: 42,
       pokemon_id: 32,
       name: 'Nidoran Male',
       nickname: 'Nidoran Male',
       favorite: -0 } ] }
```

Hmm...

mackhankins committed:
Nidoran male / female fix

mackhankins committed:
th % width suggestions

mackhankins committed:
fix for checkall not working after Nidoran fix

mackhankins committed:
moved sub-table initialization from outside click event

mackhankins committed:
Revert "moved sub-table initialization from outside click event"

This reverts commit 16c33fc68f411aa0f2b1379be8d295c00af5ef12.

mackhankins committed:
checkall exclusion

vinnymac committed:
Improved detail modal

vinnymac commented:
After pulling down your changes and fixing the merge conflicts it looks like the detail isn't working because the data structure for `pokemons` has changed to be an array of arrays. Each pokemon in pokemons used to represent a unique one, but now each pokemon in the pokemons array contains an array of unique pokemon. In order to get the modal to display all we need to do is

``` javascript
let pokemon = pokemons.pokemon.find(poke => {
    return poke.pokemon.find(pokemon => {
      return pokemon['id'] === id
  })
})
```

This won't fix all the issues though, because keys have changed places, like name and cp. I will list the changes necessary for a merge to make this more clear 👍 
- [x] Merge Conflicts
- [x] Iterate over 2D Arrays, or bind the data differently
- [x] Pass both Unique by ID and Pokemon base information to detail (poke and pokemon vars above)
- [x] Typo: prep_diplay -> prep_display
- [x] Expanding/collapsing the tables seem to make all pokemon in the list appear favorited?
- [x] Consider referring to the values in `pokemons.pokemon` as `pokemons.species` instead?
- [x] Consider not replacing the nickname and other values on the pokemon object with HTML, should probably use new keys so that we can continue to use the nickname elsewhere if we need them (like the modal)
- [x] Kill sprites until we find smaller ones to use for the table, I'll continue to use them in the modal instead

Just wanted to keep track of what I found, if I am incorrect about any of this please point it out!

vinnymac commented on js/home.js:
```diff
@@ -175,3 +298,148 @@ function addFavoriteButtonEvent () {
   })
 }
+
+function showModal (id, event) {
```
Attempt to use this showModal method instead. I got it working by making these changes on top of yours, so hopefully it will for you as well.

``` javascript
function showModal (id, event) {

  let pokemonMap = null

  pokemons.pokemon.forEach(species => {
    let pokemon = species.pokemon.find(pokemonById => {
      return pokemonById['id'] === id
    })

    if (pokemon) pokemonMap = {species, pokemon}
  })

  if (!pokemonMap) {
    console.error("No Pokemon Found to Display Detail")
    return
  }

  let $detailModal = $(detailModal)

  $detailModal.find('.modal-title').text(pokemonMap.species.name)
  $detailModal.find('.modal-body').html(detailModalBody(pokemonMap.pokemon))
  $detailModal.modal('show')
}
```

mackhankins replied:
Works perfect.  I do agree about refactoring to species though as it would be much more clear.  I'm going to try to fix that favorites bug first (great catch).

mackhankins committed:
Merge remote-tracking branch 'duhminick/master' into evolves

# Conflicts:
#	js/home.js

mackhankins committed:
Revert "Merge remote-tracking branch 'duhminick/master' into evolves"

This reverts commit 8ee4e07c66406b247beff71c44e93d2defb0991e, reversing
changes made to 2824889c9e25f50c58dd3ce90b63542e60aef34b.

mackhankins committed:
Merge remote-tracking branch 'duhminick/feature/detail' into evolves

mackhankins committed:
Typo Fix

prep_diplay -> prep_display

mackhankins committed:
Modal Fix

slight refactor for modals.

mackhankins committed:
Favorites fix

the first child expansion built the html for `favorite` causing the
second expansion to give false results.

Instead change the key to favorited for no conflict.

vinnymac commented on js/home.js:
```diff
@@ -128,6 +115,119 @@ function sortPokemonList (sorting, refresh) {
+
+      poke.checkbox = checkBox + '>'
+      poke.favorited = '<span class="favorite ' + favorite + '" id="favoriteBtn" data-pokemon-id="' + poke.id + '" data-pokemon-favorited="' + favoriteBool + '" />'
+      poke.nickname = '<a class="nickname" data-pokemon-id="' + poke.id + '">' + poke.nickname + '</a>'
```
If you rename this var to `nicknamed`, and then update line 212, I think this is good to merge 👍

mackhankins committed:
nicknamed

prep_display -> nickname renamed to nicknamed for no conflict.

mackhankins commented:
@vinnymac We should be good to go.  I want to refactor a lot of this after I get back from vacation next week or someone else can do it. Also, thanks for the help and suggestions.
- migrate to `pokemons.species` or similar
- instead of `nicknamed` maybe all the display stuff could be `td_nickname`.  Find a naming schema for display stuff.
- cleanup commented out //console.logs
